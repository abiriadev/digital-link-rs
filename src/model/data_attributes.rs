use wasm_bindgen::prelude::wasm_bindgen;

/// Maps of all possible attributes of a Digital Link
#[derive(Default, Debug, Clone)]
#[non_exhaustive]
#[wasm_bindgen(getter_with_clone, inspectable)]
pub struct DataAttributes {
	pub net_weight_vmti: Option<String>,
	pub length_vmti: Option<String>,
	pub width_vmti: Option<String>,
	pub depth_vmti: Option<String>,
	pub area_vmti: Option<String>,
	pub net_volume_vmti: Option<String>,
	pub mass_per_unit_area_vmti: Option<String>,
	pub gross_weight: Option<String>,
	pub logistic_length: Option<String>,
	pub logistic_width: Option<String>,
	pub logistic_depth: Option<String>,
	pub logistic_area: Option<String>,
	pub logistic_volume: Option<String>,
	pub processor: Option<String>,
	pub lot: Option<String>,
	pub expiry_date: Option<String>,
	pub expiry_time: Option<String>,
	pub content: Option<String>,
	pub prod_date: Option<String>,
	pub due_date: Option<String>,
	pub pack_date: Option<String>,
	pub best_before_date: Option<String>,
	pub sell_by_date: Option<String>,
	pub first_freeze_date: Option<String>,
	pub harvest_date: Option<String>,
	pub price_per_unit: Option<String>,
	pub variant: Option<String>,
	pub var_count: Option<String>,
	pub count: Option<String>,
	pub amount_pay_per_unit: Option<String>,
	pub additional_id: Option<String>,
	pub cust_part_no: Option<String>,
	pub mto_variant: Option<String>,
	pub pcn: Option<String>,
	pub secondary_serial: Option<String>,
	pub ref_to_source: Option<String>,
	pub amount: Option<String>,
	pub amount_iso: Option<String>,
	pub price: Option<String>,
	pub price_iso: Option<String>,
	pub percent_off: Option<String>,
	pub order_number: Option<String>,
	pub route: Option<String>,
	pub ship_to_loc: Option<String>,
	pub bill_to: Option<String>,
	pub purchase_from: Option<String>,
	pub ship_for_loc: Option<String>,
	pub prod_serv_loc: Option<String>,
	pub ship_to_post: Option<String>,
	pub ship_to_post_iso: Option<String>,
	pub origin: Option<String>,
	pub country_process: Option<String>,
	pub country_full_process: Option<String>,
	pub country_initial_process: Option<String>,
	pub country_disassembly: Option<String>,
	pub origin_subdivision: Option<String>,
	pub nhrn_pzn: Option<String>,
	pub nhrn_cip: Option<String>,
	pub nhrn_cn: Option<String>,
	pub nhrn_drn: Option<String>,
	pub nhrn_aim: Option<String>,
	pub nhrn_us_fda: Option<String>,
	pub nsn: Option<String>,
	pub meat_cut: Option<String>,
	pub active_potency: Option<String>,
	pub catch_area: Option<String>,
	pub fishing_gear_type: Option<String>,
	pub prod_method: Option<String>,
	pub refurb_lot: Option<String>,
	pub func_stat: Option<String>,
	pub rev_stat: Option<String>,
	pub giai_assembly: Option<String>,
	pub dimensions: Option<String>,
	pub cmt_no: Option<String>,
	pub iban: Option<String>,
	pub prod_time: Option<String>,
	pub version: Option<String>,
	pub ref_no: Option<String>,
	pub coupon_idna: Option<String>,
	pub points: Option<String>,
	pub itip_content: Option<String>,
	pub certification_ref: Option<String>,
	pub aquatic_species: Option<String>,
	pub optical_sensor: Option<String>,
	pub paperless_coupon_idna: Option<String>,
	pub internal: Option<String>,
	pub mutual: Option<String>,
	pub extension: Option<String>,
	pub ship_to_comp: Option<String>,
	pub ship_to_name: Option<String>,
	pub ship_to_add1: Option<String>,
	pub ship_to_add2: Option<String>,
	pub ship_to_sub: Option<String>,
	pub ship_to_locality: Option<String>,
	pub ship_to_reg: Option<String>,
	pub ship_to_country: Option<String>,
	pub ship_to_phone: Option<String>,
	pub rtn_to_comp: Option<String>,
	pub rtn_to_name: Option<String>,
	pub rtn_to_add1: Option<String>,
	pub rtn_to_add2: Option<String>,
	pub rtn_to_sub: Option<String>,
	pub rtn_to_loc: Option<String>,
	pub rtn_to_reg: Option<String>,
	pub rtn_to_country: Option<String>,
	pub rtn_to_post: Option<String>,
	pub rtn_to_phone: Option<String>,
	pub srv_description: Option<String>,
	pub dangerous_goods: Option<String>,
	pub auth_to_leave: Option<String>,
	pub sig_required: Option<String>,
	pub not_before_del_date: Option<String>,
	pub not_after_del_date: Option<String>,
	pub release_date: Option<String>,
	pub gtin: Option<String>,
	pub itip: Option<String>,
	pub gmn: Option<String>,
	pub cpid: Option<String>,
	pub gln: Option<String>,
	pub pay_to: Option<String>,
	pub party_gln: Option<String>,
	pub gsrnp: Option<String>,
	pub gsrn: Option<String>,
	pub gcn: Option<String>,
	pub sscc: Option<String>,
	pub gdti: Option<String>,
	pub ginc: Option<String>,
	pub gsin: Option<String>,
	pub grai: Option<String>,
	pub giai: Option<String>,
}
