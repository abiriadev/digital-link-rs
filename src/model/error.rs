use thiserror::Error;
use url::ParseError;

#[derive(Error, Debug)]
pub enum Error {
	#[error("url parsing failed: {0}")]
	UrlParseError(#[from] ParseError),
	#[error("gs1 path parsing failed: {0}")]
	Gs1PathParseError(#[from] nom::Err<String>),
}
