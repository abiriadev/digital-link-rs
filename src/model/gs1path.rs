use std::fmt::{self, Display, Formatter};

use wasm_bindgen::prelude::wasm_bindgen;

#[wasm_bindgen]
#[non_exhaustive]
#[derive(Debug)]
#[wasm_bindgen]
pub enum Gs1Path {
	Gtin {
		gtin: String,
		cpv: Option<String>,
		lot: Option<String>,
		ser: Option<String>,
	},
	Itip {
		itip: String,
		cpv: Option<String>,
		lot: Option<String>,
		ser: Option<String>,
	},
	Gmn(String),
	Cpid {
		cpid: String,
		cpsn: Option<String>,
	},
	Gln {
		gln: String,
		glnx: Option<String>,
	},
	PayTo(String),
	PartyGln(String),
	Gsrnp {
		gsrnp: String,
		srin: Option<String>,
	},
	Gsrn {
		gsrn: String,
		srin: Option<String>,
	},
	Gcn(String),
	Sscc(String),
	Gdti(String),
	Ginc(String),
	Gsin(String),
	Grai(String),
	Giai(String),
	Upui {
		gtin: String,
		tpx: String,
	},
	Eoid {
		party_gln: String,
		uic_ext: String,
	},
	Fid {
		gln: String,
		uic_ext: String,
	},
	Mid {
		giai: String,
		uic_ext: String,
	},
}

impl Gs1Path {
	fn to_code(&self) -> &'static str {
		match self {
			Gs1Path::Gtin { .. } => "01",
			Gs1Path::Itip { .. } => "8006",
			Gs1Path::Gmn(_) => "8013",
			Gs1Path::Cpid { .. } => "8010",
			Gs1Path::Gln { .. } => "414",
			Gs1Path::PayTo(_) => "415",
			Gs1Path::PartyGln(_) => "417",
			Gs1Path::Gsrnp { .. } => "8017",
			Gs1Path::Gsrn { .. } => "8018",
			Gs1Path::Gcn(_) => "255",
			Gs1Path::Sscc(_) => "00",
			Gs1Path::Gdti(_) => "253",
			Gs1Path::Ginc(_) => "401",
			Gs1Path::Gsin(_) => "402",
			Gs1Path::Grai(_) => "8003",
			Gs1Path::Giai(_) => "8004",
			Gs1Path::Upui { .. } => "01",
			Gs1Path::Eoid { .. } => "417",
			Gs1Path::Fid { .. } => "414",
			Gs1Path::Mid { .. } => "8004",
		}
	}

	fn stringify_primary(&self) -> String {
		format!("/{}/{}", self.to_code(), match self {
			Gs1Path::Gtin { gtin, .. } => gtin,
			Gs1Path::Itip { itip, .. } => itip,
			Gs1Path::Gmn(gmn) => gmn,
			Gs1Path::Cpid { cpid, .. } => cpid,
			Gs1Path::Gln { gln, .. } => gln,
			Gs1Path::PayTo(pay_to) => pay_to,
			Gs1Path::PartyGln(party_gln) => party_gln,
			Gs1Path::Gsrnp { gsrnp, .. } => gsrnp,
			Gs1Path::Gsrn { gsrn, .. } => gsrn,
			Gs1Path::Gcn(gcn) => gcn,
			Gs1Path::Sscc(sscc) => sscc,
			Gs1Path::Gdti(gdti) => gdti,
			Gs1Path::Ginc(ginc) => ginc,
			Gs1Path::Gsin(gsin) => gsin,
			Gs1Path::Grai(grai) => grai,
			Gs1Path::Giai(giai) => giai,
			Gs1Path::Upui { gtin, .. } => gtin,
			Gs1Path::Eoid { party_gln, .. } => party_gln,
			Gs1Path::Fid { gln, .. } => gln,
			Gs1Path::Mid { giai, .. } => giai,
		})
	}

	fn stringify_cpv(cpv: &Option<String>) -> String {
		cpv.clone().map_or_else(
			|| "".to_owned(),
			|cpv| format!("/22/{cpv}"),
		)
	}

	fn stringify_lot(lot: &Option<String>) -> String {
		lot.clone().map_or_else(
			|| "".to_owned(),
			|lot| format!("/10/{lot}"),
		)
	}

	fn stringify_ser(ser: &Option<String>) -> String {
		ser.clone().map_or_else(
			|| "".to_owned(),
			|ser| format!("/21/{ser}"),
		)
	}

	fn stringify_cpsn(cpsn: &Option<String>) -> String {
		cpsn.clone().map_or_else(
			|| "".to_owned(),
			|cpsn| format!("/8011/{cpsn}"),
		)
	}

	fn stringify_glnx(glnx: &Option<String>) -> String {
		glnx.clone().map_or_else(
			|| "".to_owned(),
			|glnx| format!("/254/{glnx}"),
		)
	}

	fn stringify_refno(refno: &Option<String>) -> String {
		refno.clone().map_or_else(
			|| "".to_owned(),
			|refno| format!("/8020/{refno}"),
		)
	}

	fn stringify_srin(srin: &Option<String>) -> String {
		srin.clone().map_or_else(
			|| "".to_owned(),
			|srin| format!("/8019/{srin}"),
		)
	}

	fn stringify_tpx(tpx: &String) -> String { format!("/235/{tpx}") }

	fn stringify_uic_ext(uic_ext: &String) -> String {
		format!("/7040/{uic_ext}")
	}

	fn stringify_qualifiers(&self) -> String {
		match self {
			Gs1Path::Gtin { cpv, lot, ser, .. } => format!(
				"{}{}{}",
				Self::stringify_cpv(cpv),
				Self::stringify_lot(lot),
				Self::stringify_ser(ser)
			),
			Gs1Path::Itip { cpv, lot, ser, .. } => format!(
				"{}{}{}",
				Self::stringify_cpv(cpv),
				Self::stringify_lot(lot),
				Self::stringify_ser(ser)
			),
			Gs1Path::Cpid { cpsn, .. } => Self::stringify_cpsn(cpsn),
			Gs1Path::Gln { glnx, .. } => Self::stringify_glnx(glnx),
			Gs1Path::Gsrnp { srin, .. } => Self::stringify_srin(srin),
			Gs1Path::Gsrn { srin, .. } => Self::stringify_srin(srin),
			Gs1Path::Upui { tpx, .. } => Self::stringify_tpx(tpx),
			Gs1Path::Eoid { uic_ext, .. } => Self::stringify_uic_ext(uic_ext),
			Gs1Path::Fid { uic_ext, .. } => Self::stringify_uic_ext(uic_ext),
			Gs1Path::Mid { uic_ext, .. } => Self::stringify_uic_ext(uic_ext),

			Gs1Path::Gmn(_)
			| Gs1Path::PayTo(_)
			| Gs1Path::PartyGln(_)
			| Gs1Path::Gcn(_)
			| Gs1Path::Sscc(_)
			| Gs1Path::Gdti(_)
			| Gs1Path::Ginc(_)
			| Gs1Path::Gsin(_)
			| Gs1Path::Grai(_)
			| Gs1Path::Giai(_) => "".to_owned(),
		}
	}

	fn stringify(&self) -> String {
		format!(
			"{}{}",
			self.stringify_primary(),
			self.stringify_qualifiers()
		)
	}
}

impl Display for Gs1Path {
	fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
		write!(f, "{}", self.stringify())
	}
}
