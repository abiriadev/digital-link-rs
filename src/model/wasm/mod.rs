pub mod gs1path;
