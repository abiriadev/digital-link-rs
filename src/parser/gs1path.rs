use nom::{
	branch::alt,
	combinator::opt,
	sequence::{pair, tuple},
	IResult, Parser,
};

use super::{
	primary_key::{
		cpid, gcn, gdti, giai, ginc, gln, gmn, grai, gsin, gsrn, gsrnp, gtin,
		itip, party_gln, pay_to, sscc,
	},
	qualifier::{cpsn, cpv, glnx, lot, ser, srin, tpx, uic_ext},
};
use crate::model::Gs1Path;

pub fn gs1path(i: &str) -> IResult<&str, Gs1Path> {
	alt((
		tuple((gtin, opt(cpv), opt(lot), opt(ser))).map(
			|(gtin, cpv, lot, ser)| Gs1Path::Gtin {
				gtin: gtin.to_owned(),
				cpv: cpv.map(ToOwned::to_owned),
				lot: lot.map(ToOwned::to_owned),
				ser: ser.map(ToOwned::to_owned),
			},
		),
		tuple((itip, opt(cpv), opt(lot), opt(ser))).map(
			|(itip, cpv, lot, ser)| Gs1Path::Itip {
				itip: itip.to_owned(),
				cpv: cpv.map(ToOwned::to_owned),
				lot: lot.map(ToOwned::to_owned),
				ser: ser.map(ToOwned::to_owned),
			},
		),
		gmn.map(|gmn| Gs1Path::Gmn(gmn.to_owned())),
		pair(cpid, opt(cpsn)).map(|(cpid, cpsn)| Gs1Path::Cpid {
			cpid: cpid.to_owned(),
			cpsn: cpsn.map(ToOwned::to_owned),
		}),
		pair(gln, opt(glnx)).map(|(gln, glnx)| Gs1Path::Gln {
			gln: gln.to_owned(),
			glnx: glnx.map(ToOwned::to_owned),
		}),
		pay_to.map(|pay_to| Gs1Path::PayTo(pay_to.to_owned())),
		party_gln.map(|party_gln| Gs1Path::PartyGln(party_gln.to_owned())),
		pair(gsrnp, opt(srin)).map(|(gsrnp, srin)| Gs1Path::Gsrnp {
			gsrnp: gsrnp.to_owned(),
			srin: srin.map(ToOwned::to_owned),
		}),
		pair(gsrn, opt(srin)).map(|(gsrn, srin)| Gs1Path::Gsrn {
			gsrn: gsrn.to_owned(),
			srin: srin.map(ToOwned::to_owned),
		}),
		gcn.map(|gcn| Gs1Path::Gcn(gcn.to_owned())),
		sscc.map(|sscc| Gs1Path::Sscc(sscc.to_owned())),
		gdti.map(|gdti| Gs1Path::Gdti(gdti.to_owned())),
		ginc.map(|ginc| Gs1Path::Ginc(ginc.to_owned())),
		gsin.map(|gsin| Gs1Path::Gsin(gsin.to_owned())),
		grai.map(|grai| Gs1Path::Grai(grai.to_owned())),
		giai.map(|giai| Gs1Path::Giai(giai.to_owned())),
		pair(gtin, tpx).map(|(gtin, tpx)| Gs1Path::Upui {
			gtin: gtin.to_owned(),
			tpx: tpx.to_owned(),
		}),
		pair(party_gln, uic_ext).map(|(party_gln, uic_ext)| Gs1Path::Eoid {
			party_gln: party_gln.to_owned(),
			uic_ext: uic_ext.to_owned(),
		}),
		pair(gln, uic_ext).map(|(gln, uic_ext)| Gs1Path::Fid {
			gln: gln.to_owned(),
			uic_ext: uic_ext.to_owned(),
		}),
		pair(giai, uic_ext).map(|(giai, uic_ext)| Gs1Path::Mid {
			giai: giai.to_owned(),
			uic_ext: uic_ext.to_owned(),
		}),
	))(i)
}

#[cfg(test)]
mod tests {
	use insta::assert_debug_snapshot;

	use super::*;

	#[test]
	fn should_not_accept_root_path() {
		assert_debug_snapshot!(gs1path("/"));
	}

	#[test]
	fn should_accept_spec_example_5_1() {
		assert_debug_snapshot!(gs1path("/01/09520123456788"));
	}

	#[test]
	fn should_accept_spec_example_5_2() {
		assert_debug_snapshot!(gs1path("/01/09520123456788/22/2A"));
	}

	#[test]
	fn should_accept_spec_example_5_3() {
		assert_debug_snapshot!(gs1path("/01/09520123456788/10/ABC123"));
	}

	#[test]
	fn should_accept_spec_example_5_4() {
		assert_debug_snapshot!(gs1path("/01/09520123456788/21/12345"));
	}

	#[test]
	fn should_accept_spec_example_5_5() {
		assert_debug_snapshot!(gs1path(
			"/01/09520123456788/10/ABC1/21/12345?17=180426"
		));
	}

	#[test]
	fn should_accept_spec_example_5_6() {
		assert_debug_snapshot!(gs1path(
			"/01/09520123456788?3103=000195"
		));
	}

	#[test]
	fn should_accept_spec_example_5_7() {
		assert_debug_snapshot!(gs1path(
			"/01/09520123456788?3103=000195&3922=0299&17=201225"
		));
	}

	#[test]
	fn should_accept_spec_example_5_8() {
		assert_debug_snapshot!(gs1path("/00/195201234567891232"));
	}

	#[test]
	fn should_accept_spec_example_5_9() {
		assert_debug_snapshot!(gs1path(
			"/00/195201234567891232?02=09520123456788&37=25&10=ABC12"
		));
	}

	#[test]
	fn should_accept_spec_example_5_10() {
		assert_debug_snapshot!(gs1path("/414/9520123456788"));
	}

	#[test]
	fn should_accept_spec_example_5_11() {
		assert_debug_snapshot!(gs1path(
			"/8004/9520614141234567?01=09520123456788"
		));
	}
}
