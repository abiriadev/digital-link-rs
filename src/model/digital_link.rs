use std::fmt::Display;

use nom::combinator::all_consuming;
use url::Url;

use super::{data_attributes::DataAttributes, error::Error, gs1path, Gs1Path};
use crate::parser::parse_data_attribute;

pub struct DigitalLink {
	gs1_path: Gs1Path,
	data_attributes: DataAttributes,
	base_url: String,
}

impl DigitalLink {
	pub fn try_from_str(s: &str) -> Result<DigitalLink, Error> {
		let url = Url::parse(s);

		let Ok(url) = url else { return Err(Error::UrlParseError(url.unwrap_err())); };

		let res = all_consuming(gs1path)(url.path());

		match res {
			Ok((_, gs1_path)) => {
				let mut data_attributes = DataAttributes::default();

				for query_pair in url.query_pairs() {
					parse_data_attribute(&mut data_attributes, query_pair);
				}

				Ok(Self {
					gs1_path,
					data_attributes,
					base_url: url.host_str().unwrap().to_owned(),
				})
			},
			Err(err) => Err(Error::Gs1PathParseError(
				err.map(|e| e.to_string()),
			)),
		}
	}
}

impl Display for DigitalLink {
	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
		write!(
			f,
			"http://{}{}",
			self.base_url, self.gs1_path
		)
	}
}

#[cfg(test)]
mod tests {
	use super::*;

	#[test]
	fn serialized_result_should_be_same_as_input() {
		let url =
			"http://digital-link.aipim.io/01/09520123456788/10/ABC1/21/12345";

		let dl = DigitalLink::try_from_str(url).unwrap();

		assert_eq!(url, dl.to_string());
	}
}
