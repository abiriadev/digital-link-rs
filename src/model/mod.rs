pub use data_attributes::DataAttributes;
pub use digital_link::DigitalLink;
pub use gs1path::Gs1Path;

use crate::parser::gs1path;

mod data_attributes;
mod digital_link;
mod error;
mod gs1path;
