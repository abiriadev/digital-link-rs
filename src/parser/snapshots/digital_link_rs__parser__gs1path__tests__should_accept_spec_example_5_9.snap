---
source: src/parser/gs1path.rs
expression: "gs1path(\"/00/195201234567891232?02=09520123456788&37=25&10=ABC12\")"
---
Ok(
    (
        "?02=09520123456788&37=25&10=ABC12",
        Sscc(
            "195201234567891232",
        ),
    ),
)
